feat(layout): header 컴포넌트 추가

raccoon-ccoder opened the pull request:
## 코드 변경 사항
TopBar 컴포넌트 추가
- 아이콘, 로고 포함 컴포넌트

NavigationBar 컴포넌트 추가
- 카테고리 선택 dropdown
- 카테고리별 하위 페이지 라우팅 처리

NavItem 컴포넌트 추가
- dropdown, navigation item 컴포넌트 추상화

카테고리별 하위 페이지 컴포넌트 추가
header 다크모드 CSS 적용
프로젝트 구조 변경
목록 요소 컴포넌트 리팩토링

### 변경 이유
header UI 구현

### 구현 방법
header 컴포넌트가 비대해져 기능별로 컴포넌트 분리
컴포넌트 재사용을 위한 li 태그 공통 컴포넌트 추가

## 영향 범위
화면 상단 UI 변경
navigation 클릭시 페이지 이동

## 테스트
navigation item 클릭시 페이지 이동
![스크린샷 2023-12-27 오후 10 41 47](https://github.com/f-lab-edu/stock-saga/assets/77538818/c181e821-9d69-400d-8cc9-d9f69a6bf48e)

다크모드 적용 화면
![스크린샷 2023-12-11 오후 11 31 33](https://github.com/f-lab-edu/stock-saga/assets/77538818/94076bc5-e992-4977-823f-4cb787f424ab)
![스크린샷 2023-12-11 오후 11 31 10](https://github.com/f-lab-edu/stock-saga/assets/77538818/d106faff-5637-42f0-a079-7d7c9e593842)

## 중점 리뷰 (선택)
- 중복으로 사용되는 CSS로 인해 NavItem라는 공통 컴포넌트를 만들었는데 정말 재사용성 있게 잘 만든건지 확인 받고 싶습니다. 
- header 컴포넌트가 비대해져 2개의 컴포넌트로 분리했는데 NavigationBar 컴포넌트 내부에도 코드량이 많은 것 같다는 생각도 들어 피드백 요청 드립니다.

## 반영 브랜치
feature/#6-> develop

raccoon-ccoder committed:
design(header): header  레이아웃 구성 (#6)

raccoon-ccoder committed:
Merge branch 'develop' into feature/#6

raccoon-ccoder committed:
style: prettier, tailwind css 클래스명 정렬 적용 (#6)

raccoon-ccoder committed:
design(header): 레이아웃, 모드 변환 컴포넌트 css 변경 (#6)

레이아웃 컴포넌트 내 불필요 css 삭제, header 컴포넌트 적용
선택한 모드에 따라 모드 변환 컴포넌트 icon 변경, 사이즈 변경

raccoon-ccoder committed:
refector: Header 컴포넌트를 세부 컴포넌트로 분리 (#6)

raccoon-ccoder committed:
design(header): 로고 파일 추가, header 내 로고 삽입 (#6)

raccoon-ccoder committed:
design(header): 로고 파일 수정, 다크모드 css 추가 (#6)

tailwind css 커스텀 컬러 추가
header 다크 모드 css 추가
로고 이미지 파일 추가

raccoon-ccoder committed:
feat(header): dropdown navbar, li 공통 컴포넌트 추가 (#6)

.eslintrc.json : optional props 에러 방지를 위한 default props 규칙 변경
NavItem.tsx : li 태그 공통 컴포넌트 추가
NavigationBar.tsx : navigation dropdown 추가, 현재 url과 dropdown click 이벤트 변화에 따른 하위 navigation list 변경

raccoon-ccoder committed:
fix(header): key props 관련 에러 해결 (#6)

raccoon-ccoder committed:
feat(header): 메인, 하위 nav item 클릭시 dropdown hide (#6)

raccoon-ccoder committed:
design(header): navigation item layout dark mode, hover css 추가 (#6)

raccoon-ccoder committed:
feat(router): 페이지별 컴포넌트 추가, 불필요 페이지 컴포넌트 삭제 (#6)

raccoon-ccoder committed:
refactor(header): 불필요 데이터 삭제 (#6)

raccoon-ccoder committed:
feat(header): 로고 이미지 라우팅 처리 (#6)

raccoon-ccoder committed:
Merge branch 'develop' into feature/#6

f-lab-kevin commented on app/components/layout/NavigationBar.tsx:
```diff
@@ -0,0 +1,147 @@
+import NavItem from '@/app/components/NavItem'
+
+export default function NavigationBar() {
+  const navList = [
```
상수 값들은 별도의 공간에서 관리하는게 좋지않을까요?
다양한 상수값들이 필요할 것 같은데 (스타일, 라우팅, 도메인과 관련된 값들) Next 기반으로 폴더 구조는 어떤 방식으로 작성하는게 나을까요?

f-lab-kevin reviewed (commented):
components 폴더와 layout 폴더 또한 내부에 정의되는 컴포넌트도 역할 또는 도메인별로 구분짓기 위해 폴더가 필요한것 같아요 다음주까지 다양한 Next.js Folder Structure에 대해 조사해와주시면 좋을 것 같습니다. 파일 또한 1개를 만드는 상황을 N개를 만든다 생각하시면서 확장성을 생각한다면 어떻게 지역화를 시키면서 관리할 수 있을지 고민해보시는것도 좋을것 같아요

raccoon-ccoder commented:
> components 폴더와 layout 폴더 또한 내부에 정의되는 컴포넌트도 역할 또는 도메인별로 구분짓기 위해 폴더가 필요한것 같아요 다음주까지 다양한 Next.js Folder Structure에 대해 조사해와주시면 좋을 것 같습니다. 파일 또한 1개를 만드는 상황을 N개를 만든다 생각하시면서 확장성을 생각한다면 어떻게 지역화를 시키면서 관리할 수 있을지 고민해보시는것도 좋을것 같아요

파일을 1개가 아닌 N개를 만드는 상황이라는 말씀이 잘 이해가 가지 않아 질문 드립니다.
예를 들면 버튼 컴포넌트를 만든다고 할 때 삭제 기능의 버튼, 생성 기능의 버튼, 닫기 기능의 버튼과 같이 비슷한 역할의 파일을 여러개 만드는 상황이라고 이해해도 될까요?

raccoon-ccoder committed:
refactor: 폴더 구조 변경 (#6)

메인 페이지에서 사용되는 폴더 그룹화를 위한 private folder (_home) 추가
header 역할 컴포넌트, 관련 hook 파일을 route 폴더 내부로 이동
react-query provider 파일을 provider 폴더 내부로 이동

raccoon-ccoder committed:
refactor: routes 관련 상수값 폴더 위치 변경 (#6)

각 url별 정보를 포함하는 routes 관련 상수값을 constants 폴더로 이동

raccoon-ccoder commented on app/provider/ReactQueryProvider.tsx:
기능 중심의 프로젝트 구조에서 어떤 기능과도 관련이 없는 파일이라 판단되어 `provider`폴더로 이동시켰습니다.

raccoon-ccoder commented on app/(route)/_home/components/Header.tsx:
route 폴더 내부에 메인 페이지라는 `_home`폴더를 추가했습니다.
_home 폴더 내부에는 홈페이지에 사용되는 components, hooks 관련 파일만을 관리하기에 라우팅 시스템에 고려되지 않는다는 의미를 가지는 next.js의 private folder convention을 사용했습니다.
기존 `componenets` 폴더에 있던 컴포넌트들(`Header`, `Topbar`, `NavagationBar`, `NavItem`)은 `home`이라는 기능을 담당하며 다른 route 폴더에서 사용되지 않기에 root level에 있는 componenets 폴더가 아닌 `_home/components` 폴더로 옮겼습니다.

raccoon-ccoder commented on app/(route)/_home/hooks/useLocalStorage.ts:
기존 `hooks` 폴더에 있는 `useLocalStorage` 컴포넌트 파일은 `home`기능을 담당하는 컴포넌트 내부에서만 사용되므로 `_home/hooks`폴더로 이동하였습니다.

raccoon-ccoder commented on app/(route)/_home/components/NavigationBar.tsx:
url 관련 상수값은 route 폴더에 있는 어떤 기능 폴더와도 관련이 없다 판단되어 별도의 constants 폴더로 분리했습니다.
constants 내부에 공통 역할의 파일이 다수 존재한다면 역할 기반의 서브 폴더로 관리할 예정입니다.

raccoon-ccoder commented:
@f-lab-kevin
기존 역할 중심의 폴더 구조에서 기능 중심의 폴더 구조로 변경해봤습니다.
보통 기능 위주로 작업하기에 기능 중심 구조가 코드베이스를 탐색하는 방식과 더 잘 호환된다고 판단하였기 때문입니다.
관련해서 블로그 글에 자세히 작성했습니다. [프로젝트 폴더 구조 개선 글](https://jy-beak.tistory.com/191)

기능 중심의 폴더 구조로 변경하기 위해 아래와 같은 규칙을 정의했습니다.
- (route)라는 라우터 담당 폴더 내부는 기능별로 폴더를 분리한다.
- 역할 기준 폴더는 다음과 같다 - componenets, hook, types, image, constants, provider
- 각 route 폴더 내부는 역할 기준으로 서브 폴더를 분리한다.
- 새로운 파일을 추가할 경우 해당 route > 역할 기준 폴더 중 적절한 역할 폴더 하위에 추가한다.
- 2개 이상의 route 폴더 내부에서 사용되는 파일이라면 root level의 역할 기준 폴더로 이동한다.
- root level의 역할 폴더 내부에서 공통 역할의 파일들이 다수 존재한다면 서브 폴더를 추가해 그룹화 한다.  (예 : components/CreateButton, DeleteButton → componenets/Button/CreateButton, DeleteButton)

위 규칙에 기반하여 폴더 구조를 개선하였으며 멘토님 질문에 대한 답변은 상세 코멘트로 작성했으니 확인 부탁드립니다!

raccoon-ccoder committed:
refactor(header): 목록 요소 컴포넌트 변경 (#6)

HTML 요소 측면 기준 재사용성 증가를 위해 div태그로 변경
활성화 여부에 따라 border color를 부여하는 목록 내 요소 컴포넌트 역할에 맞는 컴포넌트명, property명 변경
컴포넌트 역할에 맞지 않게 사용되는 코드 제거

raccoon-ccoder committed:
refactor(header): 목록 요소 컴포넌트 폴더, 파일명 변경 (#6)

추후 별도 route 폴더에서 재사용되는 컴포넌트로 components 폴더로 파일 위치 변경
컴포넌트명과 동일한 파일명으로 변경

raccoon-ccoder commented:
@f-lab-kevin 
기존 NavItem이라는 컴포넌트를 리팩토링했습니다.

기존 컴포넌트 문제점은 2가지로 다음과 같습니다.
1. li태그로 묶여 있어 해당 컴포넌트는 ul 태그 내부에서만 사용하는 강제성이 부여돼 재사용성이 떨어짐
2. 활성화 여부에 따라 border color를 달리하는 목록 내 요소라는 컴포넌트 역할에 맞지않는 컴포넌트명, property명 

따라서 아래와 같이 개선했습니다.
1. HTML 요소 측면에서 재사용성을 고려하기 위해 li 태그를 div 태그로 변경
2. 컴포넌트 역할에 적합한 컴포넌트명, property명 수정

raccoon-ccoder commented on app/(route)/_home/components/NavItem.tsx:
```diff
@@ -1,20 +1,19 @@
-        isCurrentPath ? 'border-amber-600' : 'border-transparent'
+    <div
+      className={`border-b-4 ${
+        isActive ? 'border-amber-600' : 'border-transparent'
```
화면 설계시 활성화된 요소라면 border color를 모두 주황색으로 스타일링하는 요소들이 많은 곳에서 사용되어 class명을 외부에서 주입하지 않고 컴포넌트 내부에 명시하였습니다.

f-lab-kevin commented:
> > components 폴더와 layout 폴더 또한 내부에 정의되는 컴포넌트도 역할 또는 도메인별로 구분짓기 위해 폴더가 필요한것 같아요 다음주까지 다양한 Next.js Folder Structure에 대해 조사해와주시면 좋을 것 같습니다. 파일 또한 1개를 만드는 상황을 N개를 만든다 생각하시면서 확장성을 생각한다면 어떻게 지역화를 시키면서 관리할 수 있을지 고민해보시는것도 좋을것 같아요
> 
> 파일을 1개가 아닌 N개를 만드는 상황이라는 말씀이 잘 이해가 가지 않아 질문 드립니다. 예를 들면 버튼 컴포넌트를 만든다고 할 때 삭제 기능의 버튼, 생성 기능의 버튼, 닫기 기능의 버튼과 같이 비슷한 역할의 파일을 여러개 만드는 상황이라고 이해해도 될까요?

기능이나 페이지 뿐만아니라 컴포넌트도 비슷한 레벨의 컴포넌트가 계속 확장하고 늘어나는 상황을 가정해보는걸 말씀드린거였어요!

f-lab-kevin reviewed (approved):
고생하셨습니다!

f-lab-kevin commented:
Commit History가 좀 더 직관적이었으면 좋겠어요
